### ISSUE TITLE
Tracking of states/properties in components within vertical-collection

### ISSUE THREAD

**Issue by patrickng** [html-next/vertical-collection#155]
When the addon is adding and removing components with nested subcomponents, I'm noticing that if we quickly go to the bottom then to the top, it's mixing up the states of each individual component instance. To give some details on the set up of this list, we have a vertical-collection of activity-card components:

```
{{#vertical-collection activities containerSelector="body" lastReached=showInfinityLoader as |activity index|}}
    {{activity-card activity=activity}}
{{/vertical-collection}}
```
In each activity-card component, amongst the many types of activities that exists, it's possible to have a message type activity that shows contents of a message sent through the application, and if it's longer than 100 characters, we set a property "fullMessageShown" to false. When a user clicks "Show more" it sets that property to true.

When I'm following closely in the inspector to see what's going on, it looks like vertical-collection is re-using the component containers (at random) that has the state of that component's instance to display new content. This behavior reflects the description on your Medium blog post going in depth on how this was implemented:

> To accomplish this, vertical-collection only ever updates the content of the components for Glimmer and Ember to rerender. The ordering of non-updated components, and their content, remains intact. The changed components are then manually moved. (This works because Glimmer only needs to maintain a stable reference to an existing DOM node, similar to the way that ember-wormhole works.)

[Source](https://medium.com/build-addepar/rendering-10-million-items-with-vertical-collection-and-ember-js-3b8a4342cc84)

Therefore, whichever container that now has the new content, the properties are preserved so using the example before, the activity component instance that had "fullMessageShown" set to true, can hold random new content and so when we scroll back up, it's possible that the activity component now is rendered in a different container so the state of "fullMessageShown" is not correct.

It doesn't seem like there's a way to do this in the options the plugin exposes, so what would be a good workaround for this if we need to keep the states/properties of the instance in sync with the content of the card?

**Comment by pzuraq** [comment]
Hey, thanks for bringing this up! This is actually a very common issue with occlusion since we must reuse as much as possible to be performant, which results in us reusing component instances entirely as you noticed. 

There are two main solutions for this right now:

1. Set the `shouldRecycle` option to false. This will cause the collection to completely teardown and recreate component instances each time, which is a bit of a performance regression. However, this is the only way to get certain functionality, like the `init` hook or `{{unbound}}` helper, to work properly since they work when the component is first created.

2. The second method is to create a proxy object that wraps each item which is passed into the collection and stores the state associated with that object:

    ```hbs
    {{#vertical-collection proxies as |proxy|}}
      <!-- access the actual item -->
      {{#if proxy.fullMessageShown}}
        {{proxy.item.fullMessage}}
      {{else}}
        {{proxy.item.short}}
      {{/if}}

      <!-- storing transitory state -->
      <button {{action (toggle proxy.fullMessageShown)}}>Toggle</button>
    {{/vertical-collection}}
    ```

    Now, doing this would create another `n` items if you just wrapped the items outright, and you would have to teardown/recreate those items each time you rendered the collection, so that's not ideal at all. You can instead create a proxy object which uses `unknownProperty` and `setUnknownProperty` to set the values on a central map of some sort, which allows you to only create as many objects as are _rendered_ at any given time. We've done something like this in the new version of Ember Table: https://github.com/Addepar/ember-table/blob/master/addon/utils/cell-proxy.js

The plan is to fully document both of these solutions for v1.0.0. We'll provide a cookbook for solution 2 which will demonstrate a full minimal example on how to create the proxy objects and use only as many as are rendered.

**Comment by gilesv** [comment]
Any plans on making `shouldRecycle` a public property? It's still not documented right now.
Thanks!

**Comment by theairdemon** [comment]
I think I ran into a similar-ish situation, and I found a workaround that works within my company's codebase, YMMV.

**TLDR: I added `did-update` to the yielded component's template, updating cached variables (specifically, a stored array of functions) periodically as the user scrolls.**

My code was similar to the initial poster's `activity-card` component, looking something like this:
```Handlebars
{{#vertical-collection items as |item|}}
    {{list-item itemId=item.id itemBool=item.bool}}
{{/vertical-collection}}
```
The bug we were seeing was very strange at first. Each `list-item` should have displayed one of two difference svgs (`svg-1` or `svg-2`) depending on the value of `item.bool`. Let's say that `svg-1` is displayed if `item.bool` is **true**, and `svg-2` is displayed if it's **false**.

On initial render of the page, everything looked fine. However, when scrolling down the page, we would occasionally see double-renders, where both `svg-1` and `svg-2` were displayed beside each other. Scrolling around quickly would make these to "jump" between different items. Important to note: the double-renders only appeared on items that had `item.bool = false`. Clicking on the incorrectly-placed `svg-1` would make all instances disappear. Refreshing the webpage would bring the issue back.

I found a temporary fix, which was very ugly:
```Handlebars
BAD CODE HERE
{{#vertical-collection items as |item|}}
    {{#if item.bool}}
        {{list-item itemId=item.id itemBool=true}}
    {{else}}
        {{list-item itemId=item.id itemBool=false}}
    {{/if}}
{{/vertical-collection}}
```
Unfortunately, this led to other issues where deleting/removing items from the list wouldn't visually update until we refreshed the page. 

The problem ended up being with the interaction between `list-item` and `vertical-collection`. As we scrolled through our items, `vertical-collection` is trying to keep performance as high as possible, and in my particular situation, that led to reusing some arrays stored in the `list-item` component.

Whenever `item.bool = true`, we create a list of arrays, and clicking on `svg-1` called a method to update and retrieve those arrays (`setArray`). If `item.bool = false`, we do not create a list of arrays, and clicking on `svg-2` does not call `setArray`. With the lazy loading of `vertical-collection`, our items with `item.bool = false` would sometimes mistakenly load with filled-out arrays. In those cases, we were passing in **false**, but the `list-item` also had an array (borrowed from another, previously-loaded `list-item`), so we would render both `svg-2` and `svg-1`.

The solution that worked for me was to call `setArrays` whenever our `itemBool` was updated:
```Handlebars
{{#vertical-collection items as |item|}}
    {{list-item itemId=item.id itemBool=item.bool}}
{{/vertical-collection}}

In list-item component template:
<div class="..." {{did-update this.setArray itemBool}}>
    {{code}}
</div>
```
As stated up top, your mileage may vary with this fix; if anyone has similar issues, I'd recommend looking into variables stored within your components. They can sometimes get cached or stuck when using `vertical-collection`, so you might just have to manually update them.

Interestingly, scrolling down the webpage doesn't activate the `did-update` that much. It's only called a few times, and the arrays rebuild themselves extremely quickly on those calls, so the performance gains from using `vertical-collection` are still maintained.
